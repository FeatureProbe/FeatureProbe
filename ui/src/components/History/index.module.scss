.history {
  height: 100%;
}

.history-title {
  display: flex;
  justify-content: space-between;
  font-family: PingFangSC-Medium;
  font-size: 18px;
  color: #212529;
  align-items: center;
  height: 52px;
  padding: 0 16px 0 20px;
}

.history-title-icon {
  font-size: 16px !important;
  cursor: pointer;
  color: #495057;
}

.modal-content {
  margin-top: 16px;

  :global {
    .CodeMirror-scroll {
      background-color: rgba(85, 110, 230, 0.04);
      font-size: 12px;
      color: rgb(120, 139, 227);
      margin-bottom: 0;
      margin-right: 0;
      padding-bottom: 0;
      overflow: auto !important;
      border-radius: 8px;
      height: 298px;
    }
  }
}

.lists {
  margin-top: 8px;
  padding-left: 20px;
  padding-bottom: 24px;
  position: relative;
  height: calc(100% - 60px);
}

.loading {
  text-align: center;
  padding-right: 20px;
  line-height: 30px;
}

.hasMore {
  text-align: center;
  padding-right: 20px;
  line-height: 30px;
  cursor: pointer;
  color: #0a70f5;
}

.no-data {
  text-align: center;
  padding-right: 20px;
  line-height: 30px;
  font-family: PingFangSC-Medium;
  font-size: 18px;
  color: #adb5bc;
}

.no-data-image {
  width: 230px;
}

.version {
  display: flex;
  cursor: pointer;
  position: relative;
}

.version-left {
  width: 2px;
  background: rgba(0, 0, 0, 0.07);
  margin: 15px 12px 0px 6px;
}

.circle {
  width: 10px;
  height: 10px;
  background: #ffffff;
  border: 2px solid rgba(0, 0, 0, 0.24);
  display: inline-block;
  border-radius: 100%;
  position: absolute;
  top: 12px;
  left: 2px;
}

.circle-selected {
  border: 2px solid #0a70f5;
}

.version-selected {
  color: #495057;
}

.version-text {
  font-size: 12px;
}

.version-text-selected {
  color: #212529;
  font-size: 12px;
}

.current {
  font-size: 12px;
  color: #0a70f5;
  margin-left: 5px;
}

.version-right {
  border-radius: 8px;
  padding: 8px;
  width: 203px;
}

.version-right-selected {
  background-color: rgba(128, 188, 248, 0.1);
}

.title {
  font-family: PingFangSC-Medium;
  font-size: 15px;
  color: #74788d;
  margin-bottom: 4px;
}

.modifyBy {
  font-size: 12px;
  color: rgba(0, 0, 0, 0.38);
  margin-bottom: 4px;
  line-height: 20px;
}

.modifyTime {
  font-size: 12px;
  color: rgba(0, 0, 0, 0.38);
  margin-bottom: 4px;
  line-height: 20px;
}

.comment {
  font-size: 12px;
  color: rgba(0, 0, 0, 0.38);
  display: -webkit-box;
  overflow: hidden;
  white-space: normal!important;
  text-overflow: ellipsis;
  word-wrap: break-word;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
}

.tooltip {
  max-width: 100px;
  word-break: break-all;
}

.version-title {
  font-family: HelveticaNeue-Medium;
  color: #74788d;
  margin-right: 2px;
}

.popup {
  opacity: 0.8 !important;
}

.version-popup {
  margin-left: 8px;
}