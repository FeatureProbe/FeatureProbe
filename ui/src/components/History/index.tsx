import { SyntheticEvent } from 'react';
import { Popup } from 'semantic-ui-react';
import { FormattedMessage } from 'react-intl';
import InfiniteScroll from 'react-infinite-scroll-component';
import TextLimit from 'components/TextLimit';
import Loading from 'components/Loading';
import classNames from 'classnames';
import dayjs from 'dayjs';
import { IVersion } from 'interfaces/targeting';
import styles from './index.module.scss';

interface IProps {
  versions: IVersion[];
  hasMore: boolean;
  latestVersion: number;
  selectedVersion: number;
  isHistoryLoading: boolean;
  loadMore(): void;
  reviewHistory(version: IVersion): void;
}

const History = (props: IProps) => {
  const { versions, hasMore, selectedVersion, latestVersion, isHistoryLoading, loadMore, reviewHistory } = props;

  return (
    <div className={styles.history}>
      <div className={styles['history-title']}>
        <span>
          <FormattedMessage id='common.history.text' />
        </span>
      </div>
      <div className={styles.lists} id='scrollableDiv'>
        {
          isHistoryLoading ? <Loading /> : (
            <InfiniteScroll
              dataLength={versions.length}
              next={loadMore}
              hasMore={hasMore}
              scrollableTarget='scrollableDiv'
              loader={
                <div className={styles.loading}>
                  <FormattedMessage id='common.loading.text' />
                </div>
              }
            >
              {
                versions.length > 0 && versions.map((item) => {
                  const clsRight = classNames(
                    styles['version-right'],
                    {
                      [styles['version-right-selected']]: item?.version === Number(selectedVersion)
                    }
                  );

                  const clsCircle = classNames(
                    styles.circle,
                    {
                      [styles['circle-selected']]: item?.version === Number(selectedVersion)
                    }
                  );

                  const clsVersion = classNames(
                    {
                      [styles['version-selected']]: item?.version === Number(selectedVersion)
                    }
                  );

                  const clsVersionText = classNames(
                    styles['version-text'],
                    {
                      [styles['version-text-selected']]: item?.version === Number(selectedVersion)
                    }
                  );

                  return (
                    <div 
                      key={item.version} 
                      className={styles.version} 
                      onClick={(e: SyntheticEvent) => {
                        e.stopPropagation();
                        if (item?.version === Number(selectedVersion)) {
                          return;
                        }
                        reviewHistory(item);
                      }}
                    >
                      <div className={styles['version-left']}>
                        <div className={clsCircle}></div>
                      </div>
                      <div className={clsRight}>
                        <div className={styles.title}>
                          <span className={clsVersion}>
                            <FormattedMessage id='common.version.uppercase.text' />:
                          </span>
                          <span className={clsVersionText}>
                            { item.version }
                          </span>
                          {
                            item.version === latestVersion && (
                              <span className={styles.current}>
                                (<FormattedMessage id='common.current.version.text' />)
                              </span>
                            )
                          }
                        </div>
                        <div className={styles.modifyBy}>
                          <span className={styles['version-title']}>
                            { item.approvalStatus === 'JUMP' ? <FormattedMessage id='approvals.skipped.by' /> : <FormattedMessage id='approvals.published.by' /> }
                          </span>
                          :
                          <Popup
                            inverted
                            className={styles.popup}
                            trigger={
                              <span>{ item.createdBy }</span>
                            }
                            content={
                              <span>
                                { item.createdBy }
                                <span className={styles['version-popup']}>{ dayjs(item.createdTime).format('YYYY-MM-DD HH:mm:ss') }</span>
                              </span>
                            }
                            position='top center'
                          />
                        </div>
                        {
                          item.approvalStatus === 'PASS' && typeof item.approvalTime === 'string' && (
                            <div className={styles.modifyBy}>
                              <span className={styles['version-title']}>
                                <FormattedMessage id='approvals.reviewed.by' />
                              </span>
                              :
                              <Popup
                                inverted
                                className={styles.popup}
                                trigger={
                                  <span>{ item.approvalBy }</span>
                                }
                                content={
                                  <span>
                                    { item.approvalBy }
                                    <span className={styles['version-popup']}>{ dayjs(item.approvalTime).format('YYYY-MM-DD HH:mm:ss') }</span>
                                  </span>
                                }
                                position='top center'
                              />
                            </div>
                          )
                        }
                        {
                          item.approvalStatus === 'JUMP' && (
                            <div className={styles.comment}>
                              <span className={styles['version-title']}>
                                <FormattedMessage id='approvals.table.header.comment' />
                              </span>
                              :
                              <Popup
                                inverted
                                style={{opacity: '0.8'}}
                                className={styles.popup}
                                trigger={
                                  <span className={styles['tooltip-text']}>{ item.approvalComment }</span>
                                }
                                content={
                                  <div className={styles.tooltip}>
                                    {item.approvalComment}
                                  </div>
                                }
                                position='top left'
                              />
                            </div>
                          )
                        }
                        {
                          item.comment && <div className={styles.comment}>
                            <span className={styles['version-title']}>
                              <FormattedMessage id='targeting.publish.modal.comment' />
                            </span>
                            :
                            <span className={styles['tooltip-text']}><TextLimit text={ item.comment } /></span>
                          </div>
                        }
                      </div>
                    </div>
                  );
                })
              }
              {
                hasMore && versions.length !== 0 && (
                  <div className={styles.hasMore} onClick={loadMore}>
                    <FormattedMessage id='targeting.view.more' />
                  </div>
                )
              }
              {
                versions.length === 0 && (
                  <div className={styles['no-data']}>
                    <img className={styles['no-data-image']} src={require('images/no-data.png')} alt='no-data' />
                    <FormattedMessage id='common.nodata.text' />
                  </div>
                )
              }
            </InfiniteScroll>
          )
        }
      </div>
    </div>
  );
};

export default History;