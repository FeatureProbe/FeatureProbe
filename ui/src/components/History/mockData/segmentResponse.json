{
  "content": [
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        }
      ],
      "version": 12,
      "createdTime": "2022-11-07T15:26:42.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    },
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        }
      ],
      "version": 11,
      "createdTime": "2022-11-07T15:23:24.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    },
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        }
      ],
      "version": 10,
      "createdTime": "2022-11-07T15:21:52.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    },
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        }
      ],
      "version": 9,
      "createdTime": "2022-11-07T15:21:15.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    },
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [
        {
          "conditions": [],
          "name": ""
        }
      ],
      "version": 8,
      "createdTime": "2022-11-07T15:18:34.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    },
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [],
      "version": 7,
      "createdTime": "2022-11-07T15:18:27.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    },
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        }
      ],
      "version": 6,
      "createdTime": "2022-11-07T15:17:51.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    },
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        }
      ],
      "version": 5,
      "createdTime": "2022-11-07T15:12:53.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    },
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        }
      ],
      "version": 4,
      "createdTime": "2022-11-07T15:10:00.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    },
    {
      "projectKey": "My_Project",
      "key": "test3",
      "comment": "",
      "disabled": false,
      "rules": [
        {
          "conditions": [],
          "name": ""
        },
        {
          "conditions": [],
          "name": ""
        }
      ],
      "version": 3,
      "createdTime": "2022-11-07T15:09:12.000+08:00",
      "createdBy": "panchongyang.p@qq.com"
    }
  ],
  "pageable": {
    "sort": {
      "unsorted": false,
      "sorted": true,
      "empty": false
    },
    "pageNumber": 0,
    "pageSize": 10,
    "offset": 0,
    "unpaged": false,
    "paged": true
  },
  "last": false,
  "totalPages": 2,
  "totalElements": 12,
  "sort": {
    "unsorted": false,
    "sorted": true,
    "empty": false
  },
  "first": true,
  "numberOfElements": 10,
  "size": 10,
  "number": 0,
  "empty": false
}