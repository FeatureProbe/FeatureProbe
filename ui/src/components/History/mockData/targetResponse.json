{
  "content": [
    {
      "projectKey": "My_First_Project",
      "environmentKey": "jianggangkey",
      "toggleKey": "json_test2",
      "comment": "test",
      "content": {
        "rules": [],
        "disabledServe": {
          "select": 0
        },
        "defaultServe": {
          "select": 0
        },
        "variations": [
          {
            "value": "{}",
            "name": "variation1",
            "description": ""
          },
          {
            "value": "{}",
            "name": "variation2",
            "description": ""
          }
        ]
      },
      "disabled": false,
      "version": 2,
      "createdTime": "2022-11-16T10:22:29.000+08:00",
      "createdBy": "Admin",
      "approvalStatus": "PASS",
      "approvalBy": "Admin",
      "approvalTime": "2022-11-16T10:22:20.000+08:00",
      "approvalComment": "test"
    },
    {
      "projectKey": "My_First_Project",
      "environmentKey": "jianggangkey",
      "toggleKey": "json_test2",
      "comment": "",
      "content": {
        "rules": [],
        "disabledServe": {
          "select": 0
        },
        "defaultServe": {},
        "variations": [
          {
            "value": "{}",
            "name": "variation1",
            "description": ""
          },
          {
            "value": "{}",
            "name": "variation2",
            "description": ""
          }
        ]
      },
      "disabled": true,
      "version": 1,
      "createdTime": "2022-11-15T17:07:07.000+08:00",
      "createdBy": "Admin"
    }
  ],
  "pageable": {
    "sort": {
      "sorted": true,
      "unsorted": false,
      "empty": false
    },
    "pageNumber": 0,
    "pageSize": 10,
    "offset": 0,
    "paged": true,
    "unpaged": false
  },
  "totalPages": 1,
  "last": true,
  "totalElements": 2,
  "sort": {
    "sorted": true,
    "unsorted": false,
    "empty": false
  },
  "first": true,
  "numberOfElements": 2,
  "size": 10,
  "number": 0,
  "empty": false
}