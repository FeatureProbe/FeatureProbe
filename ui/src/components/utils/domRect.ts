export class DOMRect {
    x: number;
    width: number;
    y: number;
    height: number;
    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
}