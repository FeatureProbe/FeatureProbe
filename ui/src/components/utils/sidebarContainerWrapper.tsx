import { SidebarContainer } from 'layout/hooks';

export default SidebarContainer.Provider;
