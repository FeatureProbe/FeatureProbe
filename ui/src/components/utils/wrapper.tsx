import en_US from '../../locales/en-US.json';
import { IntlProvider } from 'react-intl';

export const IntlWrapper: React.FC = ({ children }) => {
  return (
    <IntlProvider locale="en" messages={en_US}>{children}</IntlProvider>
  );
};
