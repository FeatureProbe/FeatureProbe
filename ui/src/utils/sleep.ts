const sleep = (delay: number) => {
  return new Promise((res) => {
    setTimeout(() => {
      res('ok');
    }, delay);
  });
};

export default sleep;
