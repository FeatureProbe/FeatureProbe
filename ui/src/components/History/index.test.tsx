import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { IntlWrapper } from 'components/utils/wrapper';
import History from '.';
import segmentResponse from './mockData/segmentResponse.json';
import targetResponse from './mockData/targetResponse.json';
import sleep from 'utils/sleep';

it('History snapshot', (done) => {
  (async () => {
    const { asFragment, rerender } = render(
      <History
        versions={[]}
        hasMore={true}
        latestVersion={0}
        loadMore={jest.fn()}
        selectedVersion={0}
        isHistoryLoading={false}
        reviewHistory={jest.fn()}
      />,
      {
        wrapper: IntlWrapper,
      }
    );

    expect(asFragment()).toMatchSnapshot();

    rerender(<History
      versions={[]}
      hasMore={true}
      latestVersion={0}
      loadMore={jest.fn()}
      selectedVersion={0}
      isHistoryLoading={true}
      reviewHistory={jest.fn()}
    />);

    expect(asFragment()).toMatchSnapshot();
    done();
  })();
});

test('History render segment data', (done) => {
  (async () => {
    const load = jest.fn().mockResolvedValue(segmentResponse);

    const mock = jest.fn();
    const { asFragment, rerender } = render(
      <History
        versions={segmentResponse.content}
        hasMore={true}
        latestVersion={segmentResponse.content[0].version}
        loadMore={load}
        selectedVersion={0}
        isHistoryLoading={false}
        reviewHistory={mock}
      />,
      {
        wrapper: IntlWrapper,
      }
    );

    expect(asFragment()).toMatchSnapshot();

    await userEvent.click(screen.getByText('Load more'));
    await sleep(0);

    expect(asFragment()).toMatchSnapshot();

    await userEvent.click(screen.getByText('7'));
    expect(mock).toBeCalledWith(expect.objectContaining({ version: 7 }));

    rerender(
      <History
        versions={segmentResponse.content}
        hasMore={true}
        latestVersion={segmentResponse.content[0].version}
        loadMore={load}
        selectedVersion={7}
        isHistoryLoading={false}
        reviewHistory={mock}
      />
    );
    await userEvent.click(screen.getByText('7'));
    expect(mock).toBeCalledTimes(1);
    done();
  })();
});

test('History render target data', (done) => {
  (async () => {
    const load = jest.fn().mockResolvedValue(targetResponse);

    const mock = jest.fn();
    const { asFragment } = render(
      <History
        versions={targetResponse.content}
        hasMore={true}
        latestVersion={targetResponse.content[0].version}
        loadMore={load}
        selectedVersion={0}
        isHistoryLoading={false}
        reviewHistory={mock}
      />,
      {
        wrapper: IntlWrapper,
      }
    );

    expect(asFragment()).toMatchSnapshot();

    done();
  })();
});


