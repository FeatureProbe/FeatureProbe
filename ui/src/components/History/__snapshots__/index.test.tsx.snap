// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History render segment data 1`] = `
<DocumentFragment>
  <div
    class="history"
  >
    <div
      class="history-title"
    >
      <span>
        History
      </span>
    </div>
    <div
      class="lists"
      id="scrollableDiv"
    >
      <div
        class="infinite-scroll-component__outerdiv"
      >
        <div
          class="infinite-scroll-component "
          style="height: auto; overflow: auto;"
        >
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  12
                </span>
                <span
                  class="current"
                >
                  (Current version)
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  11
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  10
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  9
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  8
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  7
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  6
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  5
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  4
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  3
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="hasMore"
          >
            Load more
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`History render segment data 2`] = `
<DocumentFragment>
  <div
    class="history"
  >
    <div
      class="history-title"
    >
      <span>
        History
      </span>
    </div>
    <div
      class="lists"
      id="scrollableDiv"
    >
      <div
        class="infinite-scroll-component__outerdiv"
      >
        <div
          class="infinite-scroll-component "
          style="height: auto; overflow: auto;"
        >
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  12
                </span>
                <span
                  class="current"
                >
                  (Current version)
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  11
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  10
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  9
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  8
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  7
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  6
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  5
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  4
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  3
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  panchongyang.p@qq.com
                </span>
              </div>
            </div>
          </div>
          <div
            class="hasMore"
          >
            Load more
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`History render target data 1`] = `
<DocumentFragment>
  <div
    class="history"
  >
    <div
      class="history-title"
    >
      <span>
        History
      </span>
    </div>
    <div
      class="lists"
      id="scrollableDiv"
    >
      <div
        class="infinite-scroll-component__outerdiv"
      >
        <div
          class="infinite-scroll-component "
          style="height: auto; overflow: auto;"
        >
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  2
                </span>
                <span
                  class="current"
                >
                  (Current version)
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  Admin
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Reviewed by
                </span>
                :
                <span>
                  Admin
                </span>
              </div>
              <div
                class="comment"
              >
                <span
                  class="version-title"
                >
                  Changes description
                </span>
                :
                <span
                  class="tooltip-text"
                >
                  <span
                    class="limit-str-container-w"
                    style="max-width: 100%;"
                  >
                    test
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            class="version"
          >
            <div
              class="version-left"
            >
              <div
                class="circle"
              />
            </div>
            <div
              class="version-right"
            >
              <div
                class="title"
              >
                <span
                  class=""
                >
                  Version:
                </span>
                <span
                  class="version-text"
                >
                  1
                </span>
              </div>
              <div
                class="modifyBy"
              >
                <span
                  class="version-title"
                >
                  Published by
                </span>
                :
                <span>
                  Admin
                </span>
              </div>
            </div>
          </div>
          <div
            class="hasMore"
          >
            Load more
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`History snapshot 1`] = `
<DocumentFragment>
  <div
    class="history"
  >
    <div
      class="history-title"
    >
      <span>
        History
      </span>
    </div>
    <div
      class="lists"
      id="scrollableDiv"
    >
      <div
        class="infinite-scroll-component__outerdiv"
      >
        <div
          class="infinite-scroll-component "
          style="height: auto; overflow: auto;"
        >
          <div
            class="no-data"
          >
            <img
              alt="no-data"
              class="no-data-image"
              src="no-data.png"
            />
            No Data
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`History snapshot 2`] = `
<DocumentFragment>
  <div
    class="history"
  >
    <div
      class="history-title"
    >
      <span>
        History
      </span>
    </div>
    <div
      class="lists"
      id="scrollableDiv"
    >
      <div
        class="ui active transition visible inverted dimmer"
        style="display: flex;"
      >
        <div
          class="content"
        >
          <div
            class="ui small text loader"
          >
            Loading...
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
